### ISSUE TITLE
feat(gp): collapsed latent-factor likelihood with marginalized decoder

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#197]
## Summary

Add `pyrox_gp._latent_factor`: a multi-output GP formulation in which the linear
decoder (mixing matrix) is a **random variable with a fixed `N(0, I)` prior that
is analytically marginalized**, rather than a fixed array as in the existing
`LMCKernel` / `ICMKernel` / `OILMMKernel`.

This module is pure functions only — no model class, no NumPyro sites. The model
surface is issue [C].

## Motivation

Every multi-output structure currently in `pyrox_gp` stores the mixing matrix as
a plain `Float[Array, "P Q"]` field (`_multi_output.py:189, 302, 424`). There is
no prior on it, no marginalization, and no posterior over it. Consequently:

- The `P ≫ N` regime (many outputs, few training points) has no calibrated path.
  `OILMMGPPrior` requires a *known* semi-orthogonal `W`; estimating it by point
  mass is exactly where the error budget goes when `N` is small.
- The dense path (`MultiOutputGPPrior`, `mo_gp_factor`) assembles a `(PN, PN)`
  operator — `O(P³N³)`. Infeasible for `P` in the thousands.

Collapsing in the opposite direction (integrate out `W`, point-estimate the
latents) gives a likelihood whose only factorization is `Q×Q` and whose cost is
**linear in `P`**.

Adapted from [`edstevenson/GPLFR`](https://github.com/edstevenson/GPLFR)
(Stevenson et al. 2026, arXiv:2606.06576), MIT. Attribution goes in the module
docstring.

## Math specification

Generative model, with `Z ∈ ℝ^{N×Q}` the latent GP factor values at the training
inputs and `W ∈ ℝ^{Q×P}` the decoder:

$$
W_{qj} \sim \mathcal{N}(0,1) \ \text{i.i.d.}, \qquad Y = ZW + E, \qquad E_{ij} \sim \mathcal{N}(0, \sigma^2)
$$

Marginalizing `W` column by column: for output channel `j`, `y_j = Z w_j + ε_j`
with `w_j ~ N(0, I_Q)`, so

$$
y_j \sim \mathcal{N}(0, C), \qquad C = ZZ^\top + \sigma^2 I_N
$$

**`C` is `N×N` and identical for all `P` output channels.** That is the whole
method — the outputs are conditionally i.i.d. given `Z`.

Define the `Q×Q` capacitance matrix

$$
\Psi = I_Q + \sigma^{-2} Z^\top Z
$$

Woodbury gives `log|C| = N log σ² + log|Ψ|` and
`C⁻¹ = σ⁻²I_N − σ⁻⁴ Z Ψ⁻¹ Zᵀ`, hence

$$
\begin{aligned}
\log p(Y \mid Z, \sigma) = &-\tfrac{NP}{2}\log 2\pi - \tfrac{NP}{2}\log\sigma^2 - \tfrac{P}{2}\log|\Psi| \\
&- \tfrac{1}{2\sigma^2}\|Y\|_F^2 + \tfrac{1}{2\sigma^4}\,\mathrm{tr}\!\left[(Z^\top Y)^\top \Psi^{-1} (Z^\top Y)\right]
\end{aligned}
$$

The `P` outputs enter only through `‖Y‖_F²` (scalar) and `Z^⊤Y` (`Q×P`, one GEMM).

**Decoder posterior** (conjugate, matrix normal):

$$
p(W \mid Z, Y, \sigma) = \mathcal{MN}_{Q \times P}\big(\sigma^{-2}\Psi^{-1}Z^\top Y,\ \Psi^{-1},\ I_P\big)
$$

The row covariance `Ψ⁻¹` is `Q×Q` and shared across all `P` columns; the full
`QP×QP` posterior covariance is exactly `Ψ⁻¹ ⊗ I_P` and is never materialized.

**Predictive moments.** For `z_* ~ N(m, diag(v))` independent of `W`, the exact
moments of `f = z_*ᵀW` are

$$
\mathbb{E}[f_j] = m^\top \mu_{W,\cdot j}, \qquad
\mathrm{Var}[f_j] = \underbrace{m^\top \Sigma_W m}_{\text{decoder}} + \underbrace{\sum_q v_q \mu_{W,qj}^2}_{\text{latent}} + \underbrace{\sum_q v_q \Sigma_{W,qq}}_{\text{interaction}}
$$

Terms 1 and 3 do not depend on `j`, so they are `(T,1)` columns broadcast over `P`.

## Cost

| Quantity | Cost |
|---|---|
| `Z^⊤Z` | `O(NQ²)` |
| `Z^⊤Y` | `O(NQP)` — only `P`-dependent term |
| `chol(Ψ)` | `O(Q³)` |
| **Total** | **`O(NQ² + NQP + Q³)`**, memory `O(NP + Q²)` |

No `(N,N)` matrix is formed. Compare `O(P³N³)` for the dense vec-GP path.

## Files to create / change

| File | Change |
|---|---|
| `packages/pyrox-gp/src/pyrox_gp/_latent_factor.py` | **new** — the three functions below |
| `packages/pyrox-gp/src/pyrox_gp/__init__.py` | export the three names; add a bullet to the module docstring |
| `packages/pyrox-gp/tests/gp/test_latent_factor.py` | **new** — see acceptance criteria |
| `docs/api/gp.md` | add `::: pyrox_gp.collapsed_lfr_log_prob` etc. under a new "Latent factor" heading |
| `packages/pyrox-gp/CHANGELOG.md` | do **not** hand-edit — release-please manages it |

## Reference implementation

This code is validated (see "Validation evidence" below). Match repo style:
`einx` named axes for contractions, `jaxtyping` annotations, Google-style
docstrings.

> ⚠️ **`test_docstrings.py` rejects Sphinx/RST markup.** Use `$...$` / `$$...$$`
> for math and `` `name` `` / ``[`name`][path]`` for cross-references. Never
> `` :math:`...` ``, `.. math::`, or `` :class:`...` ``.

```python
# packages/pyrox-gp/src/pyrox_gp/_latent_factor.py
"""Collapsed latent-factor regression — the decoder is marginalized, not fixed.

Unlike `pyrox_gp.LMCKernel` / `pyrox_gp.ICMKernel` /
`pyrox_gp.OILMMKernel`, which hold the mixing matrix as a concrete
array, this module treats it as a random variable with a fixed
$\\mathcal{N}(0, 1)$ prior and integrates it out analytically. The
resulting likelihood factorizes only a $Q \\times Q$ matrix and costs
$O(NQP)$ in the output dimension, so $P$ in the tens of thousands is
routine.

Adapted from the GPLFR reference implementation (Stevenson et al., 2026,
arXiv:2606.06576), MIT licensed.
"""

from __future__ import annotations

import einx
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor, cho_solve
from jaxtyping import Array, Float


def _psi_factor(
    Z: Float[Array, "N Q"],
    noise_var: Float[Array, ""],
) -> tuple[Float[Array, "Q Q"], bool]:
    """Cholesky factor of the capacitance matrix.

    The capacitance matrix is $\\Psi = I_Q + \\sigma^{-2} Z^\\top Z$. One
    factorization serves the log-density, the decoder mean, and the decoder
    covariance, so it is computed here and shared.
    """
    Q = Z.shape[1]
    gram = einx.dot("n q, n r -> q r", Z, Z)
    return cho_factor(jnp.eye(Q, dtype=Z.dtype) + gram / noise_var, lower=True)


def collapsed_lfr_log_prob(
    Y: Float[Array, "N P"],
    Z: Float[Array, "N Q"],
    noise_var: Float[Array, ""],
    *,
    jitter: float = 1e-12,
) -> Float[Array, ""]:
    """Log-likelihood with the linear decoder marginalized analytically.

    Evaluates

    $$
    p(Y \\mid Z, \\sigma^2) = \\prod_{j=1}^{P}
        \\mathcal{N}(y_j \\mid 0,\\; ZZ^\\top + \\sigma^2 I_N)
    $$

    via the Woodbury identity in the $Q \\times Q$ capacitance matrix
    $\\Psi = I_Q + \\sigma^{-2} Z^\\top Z$, so no $(N, N)$ matrix is ever
    formed. Cost is $O(NQ^2 + NQP + Q^3)$; the output dimension $P$ enters
    only through $\\lVert Y \\rVert_F^2$ and $Z^\\top Y$.

    Args:
        Y: Observations of shape ``(N, P)``. Assumed centered.
        Z: Latent factor values at the training inputs, shape ``(N, Q)``.
        noise_var: Scalar isotropic observation noise variance. Per-output
            noise is not supported — it breaks the shared-covariance identity
            the derivation rests on.
        jitter: Added to ``noise_var`` before inversion.

    Returns:
        Scalar log-likelihood.

    Examples:
        >>> import jax.numpy as jnp
        >>> Y = jnp.zeros((5, 100))
        >>> Z = jnp.ones((5, 2))
        >>> float(collapsed_lfr_log_prob(Y, Z, 0.1)) < 0.0
        True
    """
    N, P = Y.shape
    s2 = noise_var + jitter
    c, low = _psi_factor(Z, s2)
    logdet_psi = 2.0 * jnp.sum(jnp.log(jnp.diagonal(c)))
    ZTY = einx.dot("n q, n p -> q p", Z, Y)
    alpha = cho_solve((c, low), ZTY)
    quad = jnp.sum(Y * Y) / s2 - jnp.sum(ZTY * alpha) / s2**2
    return -0.5 * (
        N * P * jnp.log(2.0 * jnp.pi) + N * P * jnp.log(s2) + P * logdet_psi + quad
    )


def decoder_posterior(
    Y: Float[Array, "N P"],
    Z: Float[Array, "N Q"],
    noise_var: Float[Array, ""],
    *,
    jitter: float = 1e-12,
) -> tuple[Float[Array, "Q P"], Float[Array, "Q Q"]]:
    """Closed-form matrix-normal posterior over the decoder.

    $$
    p(W \\mid Z, Y, \\sigma) = \\mathcal{MN}_{Q \\times P}
        \\big(\\sigma^{-2}\\Psi^{-1}Z^\\top Y,\\; \\Psi^{-1},\\; I_P\\big)
    $$

    The row covariance is shared across all $P$ columns and the column
    covariance is the identity, so the full $QP \\times QP$ posterior
    covariance is exactly $\\Psi^{-1} \\otimes I_P$ and is never built.

    Args:
        Y: Observations of shape ``(N, P)``.
        Z: Latent factor values, shape ``(N, Q)``.
        noise_var: Scalar isotropic observation noise variance.
        jitter: Added to ``noise_var`` before inversion.

    Returns:
        Tuple of ``(mean, row_cov)`` with shapes ``(Q, P)`` and ``(Q, Q)``.
    """
    s2 = noise_var + jitter
    c, low = _psi_factor(Z, s2)
    mean = cho_solve((c, low), einx.dot("n q, n p -> q p", Z, Y)) / s2
    row_cov = cho_solve((c, low), jnp.eye(Z.shape[1], dtype=Z.dtype))
    return mean, row_cov


def lfr_predictive_moments(
    z_mean: Float[Array, "T Q"],
    z_var: Float[Array, "T Q"],
    mu_W: Float[Array, "Q P"],
    Sigma_W: Float[Array, "Q Q"],
    noise_var: Float[Array, ""] | None = None,
) -> tuple[Float[Array, "T P"], Float[Array, "T P"]]:
    """Exact moments of the product of two independent Gaussians.

    For $z \\sim \\mathcal{N}(m, \\mathrm{diag}(v))$ independent of
    $W \\sim \\mathcal{MN}(\\mu_W, \\Sigma_W, I_P)$, the predictive
    $f = z^\\top W$ has

    $$
    \\mathrm{Var}[f_j] = m^\\top \\Sigma_W m
        + \\sum_q v_q \\mu_{W,qj}^2
        + \\sum_q v_q \\Sigma_{W,qq}
    $$

    The first and third terms do not depend on the output index, so they are
    computed once as ``(T, 1)`` columns and broadcast over ``P``. Predictive
    variance for very wide outputs therefore costs barely more than the mean.

    Args:
        z_mean: Latent predictive means, shape ``(T, Q)``.
        z_var: Latent predictive marginal variances, shape ``(T, Q)``.
        mu_W: Decoder posterior mean, shape ``(Q, P)``.
        Sigma_W: Decoder posterior row covariance, shape ``(Q, Q)``.
        noise_var: If given, added to the variance for the observation-noise
            predictive rather than the signal predictive.

    Returns:
        Tuple of ``(mean, variance)``, both shape ``(T, P)``.
    """
    mean = einx.dot("t q, q p -> t p", z_mean, mu_W)
    decoder = jnp.sum((z_mean @ Sigma_W) * z_mean, axis=-1, keepdims=True)
    latent = einx.dot("t q, q p -> t p", z_var, mu_W**2)
    cross = z_var @ jnp.diagonal(Sigma_W)[:, None]
    var = decoder + latent + cross
    return mean, var if noise_var is None else var + noise_var
```

## Acceptance criteria

New file `packages/pyrox-gp/tests/gp/test_latent_factor.py`, with
`jax.config.update("jax_enable_x64", True)` at module top (precedent:
`tests/gp/test_markov_gp.py:27`).

- [ ] **Collapsed vs dense.** `collapsed_lfr_log_prob(Y, Z, s2)` equals
      `sum(MultivariateNormal(zeros(N), ZZᵀ+s2·I).log_prob(Y[:, j]) for j in range(P))`
      using `numpyro.distributions`. Tolerance `1e-8`.
- [ ] **Woodbury pieces, separately.** `log|C| == N·log(s2) + log|Ψ|` against
      `jnp.linalg.slogdet`, and `C⁻¹ == s2⁻¹I − s2⁻²ZΨ⁻¹Zᵀ` against
      `jnp.linalg.inv`. Tolerance `1e-10`. Pins the identity independently of
      the assembled log-density.
- [ ] **Decoder posterior.** Matches a per-column Bayesian linear regression
      (`solve(I + ZᵀZ/s2, ZᵀY[:, j]/s2)`) for the mean and `inv(I + ZᵀZ/s2)`
      for the covariance. Tolerance `1e-10`.
- [ ] **Predictive moments vs Monte Carlo.** ≥1e5 samples of
      `z ~ N(m, diag(v))`, `W ~ MN(μ_W, Σ_W, I)`; compare `f = zᵀW` moments.
      Relative tolerance `2%`. Mark `@pytest.mark.slow`.
- [ ] **Gradients.** `jax.grad(collapsed_lfr_log_prob, argnums=(1, 2))` is
      finite, and `d/d(noise_var)` matches a central finite difference to `1e-6`
      relative.
- [ ] **`jit` compatibility.** All three functions run under `jax.jit`.
- [ ] **Rotational invariance.** For orthogonal `A ∈ ℝ^{Q×Q}` (from
      `jnp.linalg.qr`), `collapsed_lfr_log_prob(Y, Z @ A, s2)` equals
      `collapsed_lfr_log_prob(Y, Z, s2)` to `1e-9`. This pins the
      identifiability structure — the `N(0,I)` prior on `W` is rotation
      invariant, so only the span of `Z` is identified. Will catch accidental
      symmetry-breaking in the `Ψ` assembly.
- [ ] **Linear scaling in `P`.** Log-density at `P = 10³, 10⁴, 10⁵` with fixed
      `N, Q`; assert wall-clock grows sub-quadratically. Mark
      `@pytest.mark.slow`.
- [ ] **Shape validation.** Mismatched `N` between `Y` and `Z` raises a clear
      `ValueError`.
- [ ] **Cross-check against `gauss_flows.GaussianPCA`** (optional test, skipped
      when the dependency is absent). See "Independent test oracle" below.

## Independent test oracle: `gauss_flows.GaussianPCA`

`gauss_flows.GaussianPCA` parameterizes `Σ = W Wᵀ + σ²I` and evaluates
`log_prob` by the same Woodbury identity and matrix-determinant lemma this issue
uses. With `W := Z` and matching `σ`, it computes exactly one column of the
collapsed likelihood:

$$
\sum_{j=1}^{P} \texttt{GaussianPCA}(W{=}Z,\ \sigma).\texttt{log\_prob}(y_j)
\;=\;
\texttt{collapsed\_lfr\_log\_prob}(Y,\ Z,\ \sigma^2)
$$

This is a **stronger** check than the dense-MVN comparison above: it is a second
Woodbury implementation written by different hands, so it catches an error in the
*derivation*, not just in the arithmetic. Keep the dense-MVN test — that is the
ground truth — and add this one alongside.

```python
import equinox as eqx
import jax, jax.numpy as jnp, jax.random as jr
import pytest

gauss_flows = pytest.importorskip("gauss_flows")


def test_matches_gaussian_pca_oracle():
    pca = gauss_flows.GaussianPCA(jr.key(0), event_shape=(N,), latent_dim=Q)
    pca = eqx.tree_at(lambda m: m.W, pca, Z)
    pca = eqx.tree_at(lambda m: m.log_sigma, pca, jnp.log(jnp.sqrt(s2)))
    oracle = jnp.sum(jax.vmap(pca.log_prob)(Y.T))
    assert jnp.allclose(oracle, collapsed_lfr_log_prob(Y, Z, s2), atol=1e-8)
```

> ⚠️ **Blocked on a `gauss_flows` bug.** `GaussianPCA.log_prob` is currently
> wrong for every `σ ≠ 1` — its Woodbury quadratic divides by `sigma2**2` where
> it should divide by `sigma2` (see the companion issue
> *"fix(distributions): correct sigma power in GaussianPCA Woodbury quadratic"*
> on `jejjohnson/gauss_flows`). Until that lands, either wait, or pin this test
> at `s2 = 1.0` with a `TODO` referencing that issue — `σ² = 1` is the one value
> where the buggy and correct expressions coincide.

Note the `vmap` keeps the cost right: `chol(Ψ)` depends only on `W` and `σ`,
which are unbatched, so JAX computes it once rather than `P` times.

## Validation evidence

The reference implementation above was executed against independent references:

| Check | Result |
|---|---|
| Collapsed vs dense per-column MVN | `9.5e-10` (x64) |
| `d/d(noise_var)` autodiff vs central difference | `6.9e-11` relative |
| Decoder mean / covariance vs brute force | `6.4e-14` / `1.6e-13` |
| Predictive moments vs 400k-sample MC | within MC noise (`0.45%`) |
| Wall clock, `P = 10³ → 10⁴ → 10⁵` | `0.155 → 0.883 → 6.996 ms` (linear) |

## Out of scope

- Model class, NumPyro sites, `numpyro.factor` registration → issue [C]
- Tempering → issue [C] (`beta`) and [D]
- Sparse / inducing-point variant. **Non-trivial**: the collapsed form needs `Z`
  at all `N` training points to build `ZᵀZ` and `ZᵀY`, so an inducing
  approximation requires a variational bound on the collapsed likelihood. Open
  research question, not a follow-up.
- Non-Gaussian likelihoods — the collapse depends on conjugacy.
- Per-output (heteroscedastic) noise — breaks the shared-`C` identity.

## Before reporting done

Per `AGENTS.md`:

```bash
uv run pytest -q                      # 0 failures
uv run --group lint ruff check .      # note: '.', not 'packages/'
uv run --group lint ruff format --check .
uv run --group typecheck ty check packages/pyrox-gp/src/pyrox_gp
```
